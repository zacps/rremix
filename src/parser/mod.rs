use pest_derive::*;

#[derive(Parser)]
#[grammar = "parser/grammar.pest"]
pub struct RemixParser;

#[cfg(test)]
mod tests {
    use crate::format_pair;
    use insta::assert_snapshot;
    use pest::Parser;
    use std::fs;
    use test_case::test_case;

    use super::*;

    #[test_case("../src/standard-lib.rem")]
    //#[test_case("ex/factorial.rem")]
    //#[test_case("ex/factorial2.rem")]
    //#[test_case("ex/primes.rem")]
    #[test_case("ex/test1.rem")]
    #[test_case("ex/test2.rem")]
    #[test_case("ex/test3.rem")]
    #[test_case("ex/test4.rem")]
    #[test_case("ex/test5.rem")]
    #[test_case("ex/test6.rem")]
    #[test_case("ex/test7.rem")]
    #[test_case("ex/test8.rem")]
    #[test_case("ex/test9.rem")]
    #[test_case("ex/test10.rem")]
    #[test_case("ex/test11.rem")]
    #[test_case("ex/test12.rem")]
    #[test_case("ex/test13.rem")]
    #[test_case("ex/test14.rem")]
    #[test_case("ex/test15.rem")]
    #[test_case("ex/test16.rem")]
    #[test_case("ex/test17.rem")]
    #[test_case("ex/test18.rem")]
    #[test_case("ex/test19.rem")]
    #[test_case("ex/test20.rem")]
    #[test_case("ex/test21.rem")]
    #[test_case("ex/test22.rem")]
    #[test_case("ex/test23.rem")]
    #[test_case("ex/test24.rem")]
    #[test_case("ex/test25.rem")]
    #[test_case("ex/test26.rem")]
    // #[test_case("ex/test27.rem")]
    #[test_case("ex/test28.rem")]
    //#[test_case("gr/drawing.rem")]
    //#[test_case("gr/arrows.rem")]
    //#[test_case("gr/squares.rem")]
    //#[test_case("gr/bounce.rem")]
    //#[test_case("ob/obj-demo.rem")]
    //#[test_case("ob/sidewindermaze.rem")]
    //#[test_case("ex/test29.rem")]
    //#[test_case("ex/test30.rem")]
    //#[test_case("ex/test31.rem")]
    //#[test_case("ex/test32.rem")]
    //#[test_case("ex/test33.rem")]
    //#[test_case("ex/test34.rem")]
    //#[test_case("ex/test35.rem")]
    //#[test_case("ex/test36.rem")]
    //#[test_case("ex/test37.rem")]
    //#[test_case("ex/test38.rem")]
    //#[test_case("ex/test39.rem")]
    fn test_parser(path: &'static str) {
        let file =
            fs::read_to_string("tests/".to_owned() + path).expect("could not find test file");
        match RemixParser::parse(Rule::program, &file) {
            Ok(mut parse) => {
                let pair = parse.next();
                assert_snapshot!(path, format_pair(&pair.unwrap(), 0, true))
            }
            Err(e) => {
                eprintln!("{e}");
                panic!("Failed to parse file {path}")
            }
        }
    }
}
