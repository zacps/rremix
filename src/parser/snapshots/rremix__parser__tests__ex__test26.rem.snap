---
source: src/parser/mod.rs
assertion_line: 72
expression: "format_pair(&pair.unwrap(), 0, true)"
---
- program
  - eol_comment: "Demonstrate joining a list of words using a separator."
  - function_definition
    - function_signature
      - name_paren > name_part > name_list > single_name_part: "join"
      - name_paren > paren > single_name_part: "list"
      - name_paren > name_part > name_list > single_name_part: "with"
      - name_paren > paren > single_name_part: "separator"
    - colon: ":"
    - function_statements
      - statement > assignment_statement
        - variable: "string"
        - colon: ":"
        - expression > unary_expression > simple_expression > literal > string: "\"\""
      - statement > expression > unary_expression > simple_expression > function_call
        - single_name_part: "for"
        - single_name_part: "each"
        - expression > unary_expression > simple_expression > variable: "word"
        - single_name_part: "in"
        - expression > unary_expression > simple_expression > variable: "list"
        - newindent_block
          - statement > expression > unary_expression > simple_expression > function_call
            - single_name_part: "append"
            - expression > unary_expression > simple_expression > variable: "word"
            - single_name_part: "to"
            - expression > unary_expression > simple_expression > variable: "string"
          - statement > expression > unary_expression > simple_expression > function_call
            - single_name_part: "if"
            - expression > unary_expression > simple_expression > function_call
              - single_name_part: "not"
              - single_name_part: "at"
              - single_name_part: "end"
              - single_name_part: "of"
              - expression > unary_expression > simple_expression > variable: "list"
            - newindent_block > statement > expression > unary_expression > simple_expression > function_call
              - single_name_part: "append"
              - expression > unary_expression > simple_expression > variable: "separator"
              - single_name_part: "to"
              - expression > unary_expression > simple_expression > variable: "string"
      - statement > expression > unary_expression > simple_expression > variable: "string"
  - statement > assignment_statement
    - variable: "words"
    - colon: ":"
    - expression > unary_expression > simple_expression > literal > list
      - list_item > expression > unary_expression > simple_expression > literal > string: "\"cat\""
      - list_item > expression > unary_expression > simple_expression > literal > string: "\"dog\""
      - list_item > expression > unary_expression > simple_expression > literal > string: "\"rat\""
      - list_item > expression > unary_expression > simple_expression > literal > string: "\"pig\""
  - statement > assignment_statement
    - variable: "result"
    - colon: ":"
    - expression > unary_expression > simple_expression > function_call
      - single_name_part: "join"
      - expression > unary_expression > simple_expression > variable: "words"
      - single_name_part: "with"
      - literal > string: "\"-\""
  - statement > expression > unary_expression > simple_expression > function_call
    - single_name_part: "showline"
    - expression > unary_expression > simple_expression > variable: "result"
  - function_definition
    - function_signature
      - name_paren > name_part > name_list > single_name_part: "from"
      - name_paren > paren > single_name_part: "list"
      - name_paren > name_part > name_list > single_name_part: "execute"
      - name_paren > paren > single_name_part: "block"
      - name_paren > name_part
        - name_list > single_name_part: "with"
        - name_list > single_name_part: "each"
      - name_paren > ref_paren > ref_name > single_name_part: "item"
    - colon: ":"
    - function_statements > statement > expression > unary_expression > simple_expression > function_call
      - single_name_part: "for"
      - single_name_part: "each"
      - expression > unary_expression > simple_expression > variable: "#item"
      - single_name_part: "in"
      - expression > unary_expression > simple_expression > variable: "list"
      - expression > unary_expression > simple_expression > variable: "block"
  - function_definition
    - function_signature
      - name_paren > name_part > name_list > single_name_part: "join2"
      - name_paren > paren > single_name_part: "list"
      - name_paren > name_part > name_list > single_name_part: "with"
      - name_paren > paren > single_name_part: "separator"
    - colon: ":"
    - function_statements
      - statement > assignment_statement
        - variable: "string"
        - colon: ":"
        - expression > unary_expression > simple_expression > literal > string: "\"\""
      - statement > expression > unary_expression > simple_expression > function_call
        - single_name_part: "from"
        - expression > unary_expression > simple_expression > variable: "list"
        - single_name_part: "execute"
        - block
          - statement > expression > unary_expression > simple_expression > function_call
            - single_name_part: "append"
            - expression > unary_expression > simple_expression > variable: "word"
            - single_name_part: "to"
            - expression > unary_expression > simple_expression > variable: "string"
          - statement > expression > unary_expression > simple_expression > function_call
            - single_name_part: "if"
            - expression > unary_expression > simple_expression > function_call
              - single_name_part: "not"
              - single_name_part: "at"
              - single_name_part: "end"
              - single_name_part: "of"
              - expression > unary_expression > simple_expression > variable: "list"
            - newindent_block > statement > expression > unary_expression > simple_expression > function_call
              - single_name_part: "append"
              - expression > unary_expression > simple_expression > variable: "separator"
              - single_name_part: "to"
              - expression > unary_expression > simple_expression > variable: "string"
        - single_name_part: "with"
        - single_name_part: "each"
        - expression > unary_expression > simple_expression > variable: "word"
      - statement > expression > unary_expression > simple_expression > variable: "string"
  - eol_comment: ".N.B. Need to fix empty strings to be copied so that the value is not kept. Done."
  - statement > assignment_statement
    - variable: "words"
    - colon: ":"
    - expression > unary_expression > simple_expression > literal > list
      - list_item > expression > unary_expression > simple_expression > literal > string: "\"cat2\""
      - list_item > expression > unary_expression > simple_expression > literal > string: "\"dog\""
      - list_item > expression > unary_expression > simple_expression > literal > string: "\"rat\""
      - list_item > expression > unary_expression > simple_expression > literal > string: "\"pig\""
  - statement > assignment_statement
    - variable: "result"
    - colon: ":"
    - expression > unary_expression > simple_expression > function_call
      - single_name_part: "join2"
      - expression > unary_expression > simple_expression > variable: "words"
      - single_name_part: "with"
      - literal > string: "\"-\""
  - statement > expression > unary_expression > simple_expression > function_call
    - single_name_part: "showline"
    - expression > unary_expression > simple_expression > variable: "result"
  - function_definition
    - function_signature
      - name_paren > name_part > name_list > single_name_part: "temp"
      - name_paren > paren > single_name_part: "more"
    - colon: ":"
    - function_statements
      - statement > assignment_statement
        - variable: "word"
        - colon: ":"
        - expression > unary_expression > simple_expression > literal > string: "\"2\""
      - statement > expression > unary_expression > simple_expression > function_call
        - single_name_part: "append"
        - expression > unary_expression > simple_expression > variable: "more"
        - single_name_part: "to"
        - expression > unary_expression > simple_expression > variable: "word"
      - statement > expression > unary_expression > simple_expression > function_call
        - single_name_part: "showline"
        - expression > unary_expression > simple_expression > variable: "word"
  - statement > expression > unary_expression > simple_expression > function_call
    - single_name_part: "temp"
    - literal > string: "\"3\""
  - statement > expression > unary_expression > simple_expression > function_call
    - single_name_part: "temp"
    - literal > string: "\"4\""
  - function_definition
    - function_signature
      - name_paren > name_part
        - name_list > single_name_part: "with"
        - name_list > single_name_part: "list"
      - name_paren > paren > single_name_part: "more"
    - colon: ":"
    - function_statements
      - statement > assignment_statement
        - variable: "list"
        - colon: ":"
        - expression > unary_expression > simple_expression > literal > list > list_item > expression > unary_expression > simple_expression > literal > string: "\"2\""
      - statement > expression > unary_expression > simple_expression > function_call
        - single_name_part: "append"
        - expression > unary_expression > simple_expression > variable: "more"
        - single_name_part: "to"
        - expression > unary_expression > simple_expression > variable: "list"
      - statement > expression > unary_expression > simple_expression > function_call
        - single_name_part: "showline"
        - expression > unary_expression > simple_expression > variable: "list"
        - single_name_part: "as"
        - single_name_part: "list"
  - statement > expression > unary_expression > simple_expression > function_call
    - single_name_part: "with"
    - single_name_part: "list"
    - literal > string: "\"3\""
  - statement > expression > unary_expression > simple_expression > function_call
    - single_name_part: "with"
    - single_name_part: "list"
    - literal > string: "\"4\""
  - EOI: ""
